File: problem-solving/_lc-1039-medium-minimum-score-triangulation-of-polygon/1-top-down.py
"""
https://leetcode.com/problems/minimum-score-triangulation-of-polygon/

Complexity
----------
Time: O(N^3)
Space: O(N^2)

Reference: https://leetcode.com/problems/minimum-score-triangulation-of-polygon/solutions/294265/python-in-depth-explanation-dp-for-beginners/
"""

class Solution:
    def minScoreTriangulation(self, values: List[int]) -> int:
        cache = {}
        def dfs(left, right):
            # Need atleast 3 points to make triangle.
            if right - left < 2:
                return 0
            if (left, right) in cache:
                return cache[(left, right)]

            cache[(left, right)] = math.inf
            for mid in range(left + 1, right):
                cache[(left, right)] = min(cache[(left, right)], values[left]*values[mid]*values[right] + dfs(left, mid) + dfs(mid, right))
            return cache[(left, right)]
        return dfs(0, len(values) - 1)

File: problem-solving/_lc-1039-medium-minimum-score-triangulation-of-polygon/1-top-down.go
/*
https://leetcode.com/problems/minimum-score-triangulation-of-polygon/

Complexity
----------
Time: O(N^3)
Space: O(N^2)

Reference: https://leetcode.com/problems/minimum-score-triangulation-of-polygon/solutions/294265/python-in-depth-explanation-dp-for-beginners/
*/

import (
    "fmt"
    "math"
)

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}

func minScoreTriangulation(values []int) int {
    var F func(int, int) int
    cache := make(map[string]int)

    F = func(left, right int) int {
        if right - left < 2 {
            return 0
        }
        key := fmt.Sprintf("%d-%d", left, right)
        if v, ok := cache[key]; ok {
            return v
        }
        cache[key] = math.MaxInt32
        for mid := left + 1; mid < right; mid++ {
            cache[key] = min(
                cache[key],
                values[left]*values[mid]*values[right] + F(left, mid) + F(mid, right))
        }
        return cache[key]
    }
    return F(0, len(values) - 1)
}
